src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncPivotBlockSelector.scala

```diff
@@ -0,0 +1,204 @@
+      .map { case (peer, _) => peer }
+
+    if (correctPeers.size >= minPeersToChoosePivotBlock) {
+
+      val (peersToAsk, waitingPeers) = correctPeers.splitAt(minPeersToChoosePivotBlock + peersToChoosePivotBlockMargin)
+
+      log.info(
+        "Trying to choose fast sync pivot block using {} peers ({} correct ones). Ask {} peers for block nr {}",
+        peersUsedToChooseTarget.size,
+        correctPeers.size,
+        peersToAsk.size,
+        expectedPivotBlock
+      )
+
+      peersToAsk.foreach(peer => obtainBlockHeaderFromPeer(peer.id, expectedPivotBlock))
+
+      val timeout = scheduler.scheduleOnce(peerResponseTimeout, self, PivotBlockTimeout)
+      context become waitingForPivotBlock(
+        peersToAsk.map(_.id).toSet,
+        waitingPeers.map(_.id),
+        expectedPivotBlock,
+        timeout,
+        Map.empty
+      )
+    } else {
+      log.info(
+        "Cannot pick pivot block. Need at least {} peers, but there are only {} which meet the criteria ({} all available at the moment). Retrying in {}",
+        minPeersToChoosePivotBlock,
+        correctPeers.size,
+        peersUsedToChooseTarget.size,
+        startRetryInterval
+      )
+      scheduleRetry(startRetryInterval)
+    }
+  }
+
+  def waitingForPivotBlock(
+      peersToAsk: Set[PeerId],
+      waitingPeers: List[PeerId],
+      pivotBlockNumber: BigInt,
+      timeout: Cancellable,
+      headers: Map[ByteString, BlockHeaderWithVotes]
+  ): Receive =
+    handleCommonMessages orElse {
+      case MessageFromPeer(blockHeaders: BlockHeaders, peerId) =>
+        peerEventBus ! Unsubscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peerId)))
+        val updatedPeersToAsk = peersToAsk - peerId
+        val targetBlockHeaderOpt =
+          if (blockHeaders.headers.size != 1) None
+          else
+            blockHeaders.headers.find(header => header.number == pivotBlockNumber)
+        targetBlockHeaderOpt match {
+          case Some(targetBlockHeader) =>
+            val newValue =
+              headers.get(targetBlockHeader.hash).map(_.vote).getOrElse(BlockHeaderWithVotes(targetBlockHeader))
+            val updatedHeaders = headers.updated(targetBlockHeader.hash, newValue)
+            votingProcess(updatedPeersToAsk, waitingPeers, pivotBlockNumber, timeout, updatedHeaders)
+          case None =>
+            blacklist(peerId, blacklistDuration, "Did not respond with pivot block header, blacklisting")
+            votingProcess(updatedPeersToAsk, waitingPeers, pivotBlockNumber, timeout, headers)
+        }
+      case PivotBlockTimeout =>
+        peersToAsk.foreach { peerId =>
+          blacklist(peerId, blacklistDuration, "Did not respond with pivot block header (timeout), blacklisting")
+        }
+        peerEventBus ! Unsubscribe()
+        log.info("Pivot block header receive timeout. Retrying in {}", startRetryInterval)
+        scheduleRetry(startRetryInterval)
+    }
+
+  private def votingProcess(
```

Comment:
why not `countingVotes` ?

From my perspective, at this stage this are states the code should express better:

`ConsensusReached`: send response and cleanup
`ElectionInProgress`: continue `runningPivotBlockElection`
`ConsensusNotReached.NoMoreVoters`: schedule a new election `startPivotBlockElection`
`ConsensusNotReached.AddNewVoter`: continue `runningPivotBlockElection` (adds extra time)

Reply:
I don't see a value of changing `votingProcess` to `countingVotes`. For me counting votes is process after the election. But it is not our case

Reply:
think is a thin line, i mean, you are checking or counting votes and defining what to do. But yeah, using the word `process` is valid too.